src/rpc/scgi_task.cc

```diff
@@ -239,36 +250,102 @@ SCgiTask::event_error() {
+    zs.next_out  = (Bytef*)output;
+    if (deflate(&zs, Z_FINISH) == Z_STREAM_ERROR)
+      return false;
+    output += chunk_size - zs.avail_out;
+  } while (zs.avail_out == 0);
+
+  // Write the header directly to the buffer. If at any point
+  // max_header_size would be exceeded, fail gracefully.
+  const std::string_view header_end("Content-Encoding: gzip\r\n\r\n");
+  const int              response_size = output - (m_buffer + max_header_size);
+  int                    header_size   = snprintf(m_buffer, max_header_size, header_template.data(), response_size);
+  if (header_size < 0 || header_end.size() > max_header_size - header_size)
+    return false;
+  std::memcpy(m_buffer + header_size, header_end.data(), header_end.size());
+  header_size += header_end.size();
+
+  // Move the response back into position right after the headers
+  std::memmove(m_buffer + header_size, m_buffer + max_header_size, response_size);
+
+  m_position   = m_buffer;
+  m_buffer_size = header_size + response_size;
+  return true;
+}
+
 bool
 SCgiTask::receive_write(const char* buffer, uint32_t length) {
   if (buffer == NULL || length > (100 << 20))
     throw torrent::internal_error("SCgiTask::receive_write(...) received bad input.");
 
-  // Need to cast due to a bug in MacOSX gcc-4.0.1.
-  if (length + 256 > std::max(m_bufferSize, (unsigned int)default_buffer_size))
-    realloc_buffer(length + 256, NULL, 0);
-
-  const auto header = m_content_type == ContentType::JSON
-                        ? "Status: 200 OK\r\nContent-Type: application/json\r\nContent-Length: %i\r\n\r\n"
-                        : "Status: 200 OK\r\nContent-Type: text/xml\r\nContent-Length: %i\r\n\r\n";
-
-  // Who ever bothers to check the return value?
-  int headerSize = sprintf(m_buffer, header, length);
-
-  m_position   = m_buffer;
-  m_bufferSize = length + headerSize;
-
-  std::memcpy(m_buffer + headerSize, buffer, length);
+  std::string header = m_content_type == ContentType::JSON
+                         ? "Status: 200 OK\r\nContent-Type: application/json\r\nContent-Length: %i\r\n"
+                         : "Status: 200 OK\r\nContent-Type: text/xml\r\nContent-Length: %i\r\n";
 
+  // Write to log prior to possible compression
   if (m_parent->log_fd() >= 0) {
     int __UNUSED result;
     // Clean up logging, this is just plain ugly...
     //    write(m_logFd, "\n---\n", sizeof("\n---\n"));
-    result = write(m_parent->log_fd(), m_buffer, m_bufferSize);
+    result = write(m_parent->log_fd(), buffer, length);
     result = write(m_parent->log_fd(), "\n---\n", sizeof("\n---\n"));
   }
 
-  lt_log_print_dump(torrent::LOG_RPC_DUMP, m_buffer, m_bufferSize, "scgi", "RPC write.", 0);
+  lt_log_print_dump(torrent::LOG_RPC_DUMP, buffer, length, "scgi", "RPC write.", 0);
+
+  // Compress the response if possible
+  if (m_client_accepts_compressed_response &&
+      gzip_enabled() &&
+      length > gzip_min_size() &&
+      gzip_compress_response(buffer, length, header)) {
```

Comment:
Don't like how a failure to compress falls back to plaintext, this should only happen if there's a serious bug so fail it completely.